File: src/main/java/net/softbell/bsh/domain/entity/NodeAction.java
package net.softbell.bsh.domain.entity;

import java.util.List;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.NamedQuery;
import javax.persistence.OneToMany;
import javax.persistence.Table;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;


/**
 * @Author : Bell(bell@softbell.net)
 * @Description : 노드 액션 엔티티
 */
@Builder
@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@Entity
@Table(name="node_action")
@NamedQuery(name="NodeAction.findAll", query="SELECT n FROM NodeAction n")
public class NodeAction
{
	@Id
	@GeneratedValue(strategy=GenerationType.IDENTITY)
	@Column(name="action_id", unique=true, nullable=false)
	private int actionId;

	@Column(nullable=false, length=50)
	private String description;

	@Column(name="enable_status", nullable=false)
	private byte enableStatus;

	@ManyToOne
	@JoinColumn(name="member_id")
	private Member member;

	@OneToMany(mappedBy="nodeAction")
	private List<NodeActionItem> nodeActionItems;

	@OneToMany(mappedBy="nodeAction")
	private List<NodeReservAction> nodeReservActions;

	@OneToMany(mappedBy="nodeActionOccur")
	private List<NodeTriggerAction> nodeTriggerActionsOccur;

	@OneToMany(mappedBy="nodeActionRestore")
	private List<NodeTriggerAction> nodeTriggerActionsRestore;

	public NodeActionItem addNodeActionItem(NodeActionItem nodeActionItem) {
		getNodeActionItems().add(nodeActionItem);
		nodeActionItem.setNodeAction(this);

		return nodeActionItem;
	}

	public NodeActionItem removeNodeActionItem(NodeActionItem nodeActionItem) {
		getNodeActionItems().remove(nodeActionItem);
		nodeActionItem.setNodeAction(null);

		return nodeActionItem;
	}

	public NodeReservAction addNodeReservAction(NodeReservAction nodeReservAction) {
		getNodeReservActions().add(nodeReservAction);
		nodeReservAction.setNodeAction(this);

		return nodeReservAction;
	}

	public NodeReservAction removeNodeReservAction(NodeReservAction nodeReservAction) {
		getNodeReservActions().remove(nodeReservAction);
		nodeReservAction.setNodeAction(null);

		return nodeReservAction;
	}

	public NodeTriggerAction addNodeTriggerActions1(NodeTriggerAction nodeTriggerActionsOccur) {
		getNodeTriggerActionsOccur().add(nodeTriggerActionsOccur);
		nodeTriggerActionsOccur.setNodeActionOccur(this);

		return nodeTriggerActionsOccur;
	}

	public NodeTriggerAction removeNodeTriggerActions1(NodeTriggerAction nodeTriggerActionsOccur) {
		getNodeTriggerActionsOccur().remove(nodeTriggerActionsOccur);
		nodeTriggerActionsOccur.setNodeActionOccur(null);

		return nodeTriggerActionsOccur;
	}

	public NodeTriggerAction addNodeTriggerActions2(NodeTriggerAction nodeTriggerActionsRestore) {
		getNodeTriggerActionsRestore().add(nodeTriggerActionsRestore);
		nodeTriggerActionsRestore.setNodeActionRestore(this);

		return nodeTriggerActionsRestore;
	}

	public NodeTriggerAction removeNodeTriggerActions2(NodeTriggerAction nodeTriggerActionsRestore) {
		getNodeTriggerActionsRestore().remove(nodeTriggerActionsRestore);
		nodeTriggerActionsRestore.setNodeActionRestore(null);

		return nodeTriggerActionsRestore;
	}
}

File: src/main/kotlin/net/softbell/bsh/domain/entity/NodeAction.kt
package net.softbell.bsh.domain.entity

import net.softbell.bsh.domain.EnableStatusRule
import java.io.Serializable
import javax.persistence.*

/**
 * @Author : Bell(bell@softbell.net)
 * @Description : 노드 액션 엔티티
 */
@Entity
@Table(name = "node_action")
@NamedQuery(name = "NodeAction.findAll", query = "SELECT n FROM NodeAction n")
class NodeAction(
        @Column(nullable = false, length = 50)
        var description: String,

        @Column(name = "enable_status", nullable = false)
        var enableStatus: EnableStatusRule,

        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name = "member_id")
        var member: Member? = null,

        @OneToMany(mappedBy = "nodeAction")
        var nodeActionItems: MutableList<NodeActionItem> = ArrayList(),

        @OneToMany(mappedBy = "nodeAction")
        var nodeReservActions: MutableList<NodeReservAction> = ArrayList(),

        @OneToMany(mappedBy = "nodeAction")
        var nodeTriggerActions: MutableList<NodeTriggerAction> = ArrayList()
) : Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "action_id", unique = true, nullable = false)
    var actionId: Long = 0

    fun addNodeActionItem(nodeActionItem: NodeActionItem): NodeActionItem? {
        nodeActionItems.add(nodeActionItem)
        nodeActionItem.nodeAction = this
        return nodeActionItem
    }

//    fun removeNodeActionItem(nodeActionItem: NodeActionItem): NodeActionItem? {
//        nodeActionItems.remove(nodeActionItem)
//        nodeActionItem.nodeAction = null
//        return nodeActionItem
//    }

    fun addNodeReservAction(nodeReservAction: NodeReservAction): NodeReservAction? {
        nodeReservActions.add(nodeReservAction)
        nodeReservAction.nodeAction = this
        return nodeReservAction
    }

//    fun removeNodeReservAction(nodeReservAction: NodeReservAction): NodeReservAction? {
//        nodeReservActions.remove(nodeReservAction)
//        nodeReservAction.nodeAction = null
//        return nodeReservAction
//    }

    fun addNodeTriggerActions(nodeTriggerAction: NodeTriggerAction): NodeTriggerAction? {
        nodeTriggerActions.add(nodeTriggerAction)
        nodeTriggerAction.nodeAction = this
        return nodeTriggerAction
    }

//    fun removeNodeTriggerActions(nodeTriggerAction: NodeTriggerAction): NodeTriggerAction? {
//        nodeTriggerActions.remove(nodeTriggerAction)
//        nodeTriggerAction.nodeAction = null
//        return nodeTriggerAction
//    }

    companion object {
        private const val serialVersionUID = 1L
    }
}